### PR TITLE
pkg/ansible - update status to include failure message on the status.

### PR DESCRIPTION
**Pull request by shawn-hurley**
**Description of the change:**
Updates the ansible operator status to align better [conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#typical-status-properties)

**Motivation for the change:**
* Adding failure message to the user to see why the operator was unable to complete
* Better align base operator to fit in with the conventions.

### CODE REVIEW

File: pkg/ansible/controller/status/types.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	Changed          int                `json:"changed"`
+	Skipped          int                `json:"skipped"`
+	Failures         int                `json:"failures"`
+	TimeOfCompletion eventapi.EventTime `json:"completion"`
+}
+
+// NewAnsibleStatusFromStatusJobEvent - creates a Ansible status from job event.
+func NewAnsibleStatusFromStatusJobEvent(je eventapi.StatusJobEvent) *AnsibleStatus {
+	// ok events.
+	o := 0
```
Comment (mhrivnak) [inline pkg/ansible/controller/status/types.go]: What do you think about just creating the `AnsibleStatus` here, letting the values default to 0, and then set them in the code below as appropriate?

File: pkg/ansible/controller/status/types.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+		Failures:         failures,
+		TimeOfCompletion: je.Created,
+	}
+}
+
+// NewAnsibleStatusFromMap - creates a Ansible status from a job event.
+func NewAnsibleStatusFromMap(sm map[string]interface{}) *AnsibleStatus {
+	//Create Old top level status
+	// ok events.
+	o := 0
```
Comment (mhrivnak) [inline pkg/ansible/controller/status/types.go]: same here

File: pkg/ansible/controller/status/types.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+package status
+
+import (
+	"time"
+
+	"github.com/operator-framework/operator-sdk/pkg/ansible/runner/eventapi"
+	"github.com/sirupsen/logrus"
+	"k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (mhrivnak) [inline pkg/ansible/controller/status/types.go]: suggest aliasing this as `metav1` for consistency with other places it gets used.

File: pkg/ansible/controller/status/utils.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		Status:             status,
+		LastTransitionTime: metav1.Now(),
+		Reason:             reason,
+		Message:            message,
+		AnsibleStatus:      ansibleStatus,
+	}
+}
+
+// GetCondition returns the condition with the provided type.
+func GetCondition(status Status, condType ConditionType) *Condition {
```
Comment (mhrivnak) [inline pkg/ansible/controller/status/utils.go]: What do you think of making this, plus Set and Remove below, methods on Status?

File: pkg/ansible/runner/eventapi/types.go [hunk 5]
```diff
@@ -20,6 +20,29 @@ import (
 	"time"
 )
 
+const (
+	// Ansible Events
+
+	// EventPlaybookOnTaskStart - playbook is starting to running.
```
Comment (mhrivnak) [inline pkg/ansible/runner/eventapi/types.go]: starting to running?

File: pkg/ansible/runner/eventapi/types.go [hunk 6]
```diff
@@ -20,6 +20,29 @@ import (
...
+	EventPlaybookOnTaskStart = "playbook_on_task_start"
+	// EventRunnerOnOk - task finished with ok status.
+	EventRunnerOnOk = "runner_on_ok"
+	// EventRunnerOnFailed - task finished with failed staus.
```
Comment (mhrivnak) [inline pkg/ansible/runner/eventapi/types.go]: s/staus/status/

File: pkg/ansible/runner/eventapi/types.go [hunk 7]
```diff
@@ -74,3 +97,31 @@ type StatsEventData struct {
 	Failures     map[string]int `json:"failures"`
 	Skipped      map[string]int `json:"skipped"`
 }
+
+// FailureMessages - failure messages from the event api
+type FailureMessages []string
+
+// GetMessages - retrieves the messages as a string
+func (f FailureMessages) GetMessages() string {
```
Comment (mhrivnak) [inline pkg/ansible/runner/eventapi/types.go]: I might be missing something, but could this whole method be replaced with `strings.Join(f, "\n")` ?

File: pkg/ansible/runner/eventapi/types.go [hunk 8]
```diff
@@ -74,3 +97,31 @@ type StatsEventData struct {
...
+		if i > 0 {
+			sb.Write([]byte("\n"))
+		}
+		sb.Write([]byte(msg))
+	}
+	return sb.String()
+}
+
+// GetFailedPlaybookMessage - get the failure message from res.msg
+func GetFailedPlaybookMessage(je JobEvent) string {
```
Comment (mhrivnak) [inline pkg/ansible/runner/eventapi/types.go]: Should this be a method on `JobEvent` ?

File: pkg/ansible/controller/reconcile.go [hunk 9]
```diff
@@ -96,34 +101,32 @@ func (r *AnsibleOperatorReconciler) Reconcile(request reconcile.Request) (reconc
...
-		reconcileResult.Requeue = true
-		return reconcileResult, nil
 	}
-	status := u.Object["status"]
-	_, ok = status.(map[string]interface{})
-	if !ok {
-		logrus.Debugf("status was not found")
-		u.Object["status"] = map[string]interface{}{}
+	statusInterface := u.Object["status"]
+	statusMap, _ := statusInterface.(map[string]interface{})
```
Comment (hasbro17) [inline pkg/ansible/controller/reconcile.go]: We're no longer need to check if `Object["status"]` exists or is of type `map[string]interface{}`?

File: pkg/ansible/controller/reconcile.go [hunk 10]
```diff
@@ -185,31 +185,42 @@ func (r *AnsibleOperatorReconciler) Reconcile(request reconcile.Request) (reconc
...
+			strings.Join(failureMessages, "\n"),
+		)
+		ansiblestatus.SetCondition(&crStatus, *c)
+	} else {
+		c := ansiblestatus.NewCondition(
+			ansiblestatus.RunningConditionType,
+			v1.ConditionTrue,
+			ansibleStatus,
+			ansiblestatus.SuccessfulReason,
+			ansiblestatus.SuccessfulMessage,
```
Comment (chancez) [inline pkg/ansible/controller/reconcile.go]: Would it be useful to record what you did to some degree in the message? Or perhaps you just want that to be in the logs/events?

File: pkg/ansible/controller/status/utils.go (hunk 4 above)
Comment (chancez) [reply pkg/ansible/controller/status/utils.go]: Upstream generally avoids adding methods to the API types, so I think it's generally preferable to use functions. Plus, everything they're messing with is public, so there's not really a reason for it to be a method.

File: pkg/ansible/controller/reconcile.go [hunk 11]
```diff
@@ -96,34 +100,28 @@ func (r *AnsibleOperatorReconciler) Reconcile(request reconcile.Request) (reconc
...
-		logrus.Debugf("status was not found")
-		u.Object["status"] = map[string]interface{}{}
+	statusInterface := u.Object["status"]
+	statusMap, _ := statusInterface.(map[string]interface{})
+	crStatus := ansiblestatus.CreateStatusFromMap(statusMap)
+
+	// If the no current status add that we are working on this resourece.
+	errCond := ansiblestatus.GetCondition(crStatus, ansiblestatus.FailureConditionType)
+	succCond := ansiblestatus.GetCondition(crStatus, ansiblestatus.RunningConditionType)
+	if (errCond == nil && succCond == nil) || (succCond != nil && succCond.Reason != ansiblestatus.SuccessfulReason) {
```
[the remark this answers is not in the document]
Comment (mhrivnak) [reply pkg/ansible/controller/reconcile.go]: :+1:

**Pull request merged**